web/shared/src/typeahead.ts

```diff
@@ -48,14 +48,41 @@ export function remove_diacritics(s: string): string {
     return s.normalize("NFKD").replace(unicode_marks, "");
 }
 
+function upper_bound_lexicographically(prefix: string, words: string[]): number {
+    let left = 0;
+    let right = words.length;
+    let found = false;
+    while (left < right) {
+        const mid = Math.floor((left + right) / 2);
+        if (words[mid].startsWith(prefix)) {
+            left = mid + 1;
+            found = true;
+        } else if (words[mid] < prefix) {
+            left = mid + 1;
+        } else {
+            right = mid;
+        }
+    }
+    if (found) {
+        return left;
+    }
+    return 0;
+}
+
+function strcmp(a: string, b: string): number {
+    return a < b ? -1 : a > b ? 1 : 0;
+}
```

Comment:
This is JavaScript's default behavior on strings, and doesn't need an explicit comparator.

Reply:
In `util.ts` I saw an identical function which is used since the default "String.localeCompare is really slow". However, doing some more research I realise `localeCompare` is only used when there are accented characters, and since we remove diacritics from `source_str` the default sort shouldn't be a problem. We don't remove diacritics from `query` though... Maybe we should?

Reply:
The default behavior of `sort` doesn't involve `String.localeCompare`.

MDN is a great reference resource for looking up exactly how some feature of JS (or CSS or HTML) is supposed to work:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#description
> If `compareFn` is not supplied, all non-`undefined` array elements are sorted by converting them to strings and comparing strings in UTF-16 code units order. For example, "banana" comes before "cherry". In a numeric sort, 9 comes before 80, but because numbers are converted to strings, "80" comes before "9" in the Unicode order.

So that comparison is the same as what the `<` operator does if you already have strings.



> We don't remove diacritics from `query` though... Maybe we should?

Hmm, yeah, if we're removing them from the text then we should also remove them from the query — that'd be a good fix. Should be its own commit, and perhaps its own PR.

Reply:
> MDN is a great reference resource for looking up exactly how some feature of JS (or CSS or HTML) is supposed to work

(What I generally do is just google what I'm looking for with the term "mdn" added. So e.g. https://www.google.com/search?q=mdn+sort goes straight to the page I linked above.)

Reply:
> Hmm, yeah, if we're removing them from the text then we should also remove them from the query — that'd be a good fix. Should be its own commit, and perhaps its own PR.

Have fixed this in a separate commit!